### DISCUSSION TITLE
Intercept gRPC support

### DISCUSSION THREAD

**Discussion by jstorm31** [cypress-io/cypress#19085] [Questions and Help]
We have Cypress tests for a web app with a gRPC backend API and would like to intercept API calls with `cy.intercept`. The interception works fine, but I'm struggling with defining response body so our grpc client accepts it. We use [grpc-web](https://github.com/improbable-eng/grpc-web) client library.

There is [cypress-protobuf](https://github.com/NoriSte/cypress-protobuf) plugin, but doesn't work out with neither our app nor `cy.intercept`. Therefore, we create the grpc messages directly and serialize it to binary representation.

```js
import { SendFeedbackEmailResponse } from '../../grpc-client/v1/user_pb';

const message = new SendFeedbackEmailResponse();
message.setStatus(1);
const body = message.serializeBinary(); // Returns UInt8Array

cy.intercept('POST', `${Cypress.env('API_URL')}/SendFeedbackEmail`, {
   statusCode: 200,
   headers: {
      'content-type': 'application/grpc-web+proto',
   },
    body,
});
```

As I said, the intercept works fine, but the app fails on grpc client failing to decode the response. That implies the issue is in a combination of correct `body` representation and maybe `content-type`.

I wonder whether there are any folks that successfully use intercepts for grpc API calls or the Cypress team plans any built-in support for it. Thx 🙌

**Comment by seanmcquaid** [comment]
Curious here but did you ever end up solving this?? I am running into the same thing currently!

**Reply by jstorm31** [reply]
Well, no 😅. At the and we discoverd we would have to dig deep into the [improbable grpc-web](https://github.com/improbable-eng/grpc-web) library we use and define the response format for `intercept` exactly the same way, so the application can understand it, which was not worth the time for our use case.

It may also work better for a different [gRPC web client](https://github.com/grpc/grpc-web).

At our company, we plan to use gRPC more often in the future, which could lead to creating a plugin to Cypress, but as I said, it isn't worth the effort at the moment for us. 🤷‍♂️

**Reply by seanmcquaid** [reply]
It's all good, I figured I would at least ask! Unfortunately we are using grpc-web and we're running into the exact same issue currently at my company. We're also planning to use gRPC a lot more in the future so I will keep you posted if we end up going down a similar path and creating a plugin for this!

**Reply by Parsa-Sedigh** [reply]
@seanmcquaid did you make it work? I want to stub some of my grpc responses with a fixture, but I can't decode the response to decide whether that request should be stubbed or not.

**Comment by AmitLevinsky** [comment]
I'm also curious if you found a solution for your question? I'm working on E2E tests with mock data. 
I managed to intercept the grpc request with `cy.intercept` and req.reply with headers and body but the response.message in the generated TS code is always **null** not matter what body is set in the req.reply of cy.intercept
we also using importable-grpc-web 
Any thoughts why i'm getting response.message that is null all the time?

**Reply by jstorm31** [reply]
Nope, I didn't find any solution.
